File: client/python/mesh/mpc/urn.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#

from itertools import repeat
from typing import List

import mesh.log as log

MESH_DOMAIN = "trustbe.cn"
LOCAL_NODE_ID = "LX0000000000000"
LOCAL_INST_ID = "JG0000000000000000"


class URN:
    """
    Like:

    create.tenant.omega . 0001 . json . http2 . lx000001 . trustbe.net
    -------------------   ----   ----   -----   --------   -----------
    name            flag   codec  proto     node       domain

    """

    @staticmethod
    def parse(urn: str) -> "URN":
        instance = URN()
        if urn is None or urn.isspace():
            log.warn(f'Unresolved urn {urn}')
            return instance
        names = URN.as_array(urn)
        if len(names) < 5:
            log.warn(f'Unresolved urn {urn}')
            return instance
        names.reverse()
        instance.name = "{}.{}".format(names[1], names[0])
        instance.node_id = names[2]
        instance.flag = URNFlag.parse(names[3])
        instance.name = ".".join(names[4:])
        return instance

    def __init__(self):
        self.domain = MESH_DOMAIN
        self.node_id = ""
        self.flag = URNFlag()
        self.name = ""

    def __str__(self):
        return self.string()

    def match_node(self, node_id: str) -> bool:
        return self.node_id.lower() == node_id.lower()

    def string(self):
        urn = self.as_array(self.name)
        urn.reverse()
        urn.append(self.flag.string())
        urn.append(self.node_id.lower())
        if self.domain.isspace():
            urn.append(MESH_DOMAIN)
        else:
            urn.append(self.domain.lower())
        return ".".join(urn)

    @staticmethod
    def as_array(urn: str) -> List[str]:
        pairs = urn.split(".")
        names = []
        buff = []
        for pair in pairs:
            if pair and pair.startswith("${"):
                buff.append(pair)
                buff.append('.')
                continue
            if pair and pair.endswith("}"):
                buff.append(pair)
                names.append(''.join(buff))
                buff.clear()
                continue
            names.append(pair)
        return names

    @staticmethod
    def loc_urn(name: str) -> str:
        return URN.any_urn(name, LOCAL_NODE_ID)

    @staticmethod
    def any_urn(name: str, node_id: str) -> str:
        urn = URN()
        urn.domain = MESH_DOMAIN
        urn.node_id = node_id
        urn.flag = URNFlag.parse("0001000000000000000000000000000000000")
        urn.name = name
        return urn.string()


class URNFlag:

    @staticmethod
    def parse(value: str):
        flags = URNFlag()
        if value is None or value.isspace():
            log.error(f'Unresolved Flag {value}')
            return flags
        flags.v = value[0:2]
        flags.proto = value[2:4]
        flags.codec = value[4:6]
        flags.version = URNFlag.reduce(value[6:12], 2)
        flags.zone = value[12:14]
        flags.cluster = value[14:16]
        flags.cell = value[16:18]
        flags.group = value[18:20]
        flags.address = URNFlag.reduce(value[20:32], 3)
        flags.port = URNFlag.reduce(value[32:37], 5)
        return flags

    @staticmethod
    def reduce(value: str, length: int):
        buff = []
        for index, code in enumerate(value):
            if (index + length) % length != 0:
                continue
            has_none_zero = True
            for offset in range(index, index + length):
                if value[offset] != '0':
                    has_none_zero = False
                if not has_none_zero:
                    buff.append(value[offset])
            if has_none_zero:
                buff.append('0')
            buff.append('.')

        return ''.join(buff[0:len(buff) - 1])

    def __init__(self):
        self.v = '00'  # #                 2      00
        self.proto = '00'  # #             2      00
        self.codec = '00'  # #             2      00
        self.version = '000000'  # #       6      000000
        self.zone = '00'  # #              2      00
        self.cluster = '00'  # #           2      00
        self.cell = '00'  # #              2      00
        self.group = '00'  # #             2      00
        self.address = '000000000000'  # # 12     000000000000
        self.port = '00080'  # #           5      00080   Max port 65535

    @staticmethod
    def padding(v: str, length: int) -> str:
        value = v.replace(".", "")
        if len(value) == length:
            return value
        if len(value) < length:
            return f'{"".join(repeat("0", length - len(v)))}{value}'
        return v[0:length]

    def padding_chain(self, v: str, length: int, size: int) -> str:
        def chain(values: [str]) -> str:
            chars = []
            for frag in values:
                chars.append(self.padding(frag, length))
            return "".join(chars)

        frags = v.split(".")
        minimal = len(frags)
        if len(frags) == size:
            return chain(frags)
        if len(frags) < size:
            for _ in range(0, size - minimal):
                frags.append("")
            return chain(frags)
        return chain(frags[0:size])

    def string(self):
        return "{}{}{}{}{}{}{}{}{}{}".format(
            self.padding(self.v, 2),
            self.padding(self.proto, 2),
            self.padding(self.codec, 2),
            self.padding_chain(self.version, 2, 3),
            self.padding(self.zone, 2),
            self.padding(self.cluster, 2),
            self.padding(self.cell, 2),
            self.padding(self.group, 2),
            self.padding_chain(self.address, 3, 4),
            self.padding(self.port, 5)
        )

    def reset(self, proto: str, codec: str, version: str, zone: str, cluster: str, cell: str, group: str, address: str,
              port: str) -> "URNFlag":
        self.proto = proto
        self.codec = codec
        self.version = version
        self.zone = zone
        self.cluster = cluster
        self.cell = cell
        self.group = group
        self.address = address
        self.port = port
        return self

File: client/javascript/src/kinds/urn.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */


import Tool from "@/tool/tool";

export class URN {
    public static CN = "trustbe";
    public static MESH_DOMAIN = this.CN + ".cn";
    public static LocalNodeId = "LX0000000000000";
    public static LocalInstId = "JG0000000000000000";
    public domain: string = "";
    public nodeId: string = "";
    public flag: URNFlag = new URNFlag();
    public name: string = "";

    public static from(urn: string): URN {
        const name = new URN();
        if (!urn) {
            console.log(`Unresolved urn ${urn}`)
            return name;
        }
        const names = name.asArray(urn);
        if (names.length < 5) {
            console.log(`Unresolved urn ${urn}`)
            return name;
        }
        names.reverse();
        name.domain = `${names.splice(1, 1).join("")}.${names.splice(0, 1).join("")}`;
        name.nodeId = names.splice(0, 1).join("");
        name.flag = URNFlag.from(names.splice(0, 1).join(""));
        name.name = names.join(".");
        return name;
    }

    public static urn(name: string): string {
        const urn = new URN();
        urn.nodeId = this.LocalNodeId;
        urn.flag = URNFlag.from("0001000000000000000000000000000000000");
        urn.name = name;
        urn.domain = this.MESH_DOMAIN;
        return urn.toString();
    }

    toString(): string {
        const urn = [];
        const names = this.asArray(this.name);
        names.reverse();
        names.filter(x => x).forEach(x => urn.push(x));
        urn.push(this.flag.toString());
        urn.push(this.nodeId);
        urn.push(Tool.anyone(this.domain, URN.MESH_DOMAIN));
        return urn.join(".");
    }

    private asArray(text: string): string[] {
        const pairs = text.split(".");
        const names = [];
        const buffer = [];
        for (let pair of pairs) {
            if (pair.startsWith("${")) {
                buffer.push(pair);
                buffer.push('.');
                continue
            }
            if (pair.endsWith("}")) {
                buffer.push(pair);
                names.push(buffer.join(""))
                buffer.splice(0, buffer.length);
                continue
            }
            names.push(pair)
        }
        return names;
    }

    matchName(name: string): boolean {
        if (!name) {
            return false;
        }
        return name.endsWith(".*") && name.startsWith(name.replace(".*", ""));
    }

    resetFlag(proto: string, codec: string): URN {
        this.flag.proto = proto;
        this.flag.codec = codec;
        return this;
    }
}

export class URNFlag {

    public v: string = "00";
    public proto: string = "00";
    public codec: string = "00";
    public version: string = "000000";
    public zone: string = "00";
    public cluster: string = "00";
    public cell: string = "00";
    public group: string = "00";
    public address: string = "000000000000";
    public port: string = "00000";

    public static from(value: string): URNFlag {
        if (!value) {
            return new URNFlag();
        }
        const flag = new URNFlag();
        flag.v = Tool.substring(value, 0, 2);
        flag.proto = Tool.substring(value, 2, 2);
        flag.codec = Tool.substring(value, 4, 2);
        flag.version = this.reduce(Tool.substring(value, 6, 6), 2);
        flag.zone = Tool.substring(value, 12, 2);
        flag.cluster = Tool.substring(value, 14, 2);
        flag.cell = Tool.substring(value, 16, 2);
        flag.group = Tool.substring(value, 18, 2);
        flag.address = this.reduce(Tool.substring(value, 20, 12), 3);
        flag.port = this.reduce(Tool.substring(value, 32, 5), 5);
        return flag;
    }

    toString(): string {
        const v = this.padding(Tool.anyone(this.v, ""), 2);
        const proto = this.padding(Tool.anyone(this.proto, ""), 2);
        const codec = this.padding(Tool.anyone(this.codec, ""), 2);
        const version = this.paddingChain(Tool.anyone(this.version, ""), 2, 3);
        const zone = this.padding(Tool.anyone(this.zone, ""), 2);
        const cluster = this.padding(Tool.anyone(this.cluster, ""), 2);
        const cell = this.padding(Tool.anyone(this.cell, ""), 2);
        const group = this.padding(Tool.anyone(this.group, ""), 2);
        const address = this.paddingChain(Tool.anyone(this.address, ""), 3, 4);
        const port = this.padding(Tool.anyone(this.port, ""), 5);
        return `${v}${proto}${codec}${version}${zone}${cluster}${cell}${group}${address}${port}`;
    }

    private padding(v: string, length: number): string {
        const value = v.replace(".", "");
        if (value.length == length) {
            return value;
        }
        if (value.length < length) {
            return Tool.repeat('0', length - v.length) + value;
        }
        return v.substring(0, length);
    }

    private paddingChain(v: string, length: number, size: number): string {
        const array = v.split("[.*]");
        const frags = array.map(x => parseInt(x) ? x : "");
        if (frags.length == size) {
            return frags.map(x => this.padding(x, length)).join("")
        }
        if (frags.length < size) {
            for (let index = 0; index < size - array.length; index++) {
                frags.push("");
            }
            return frags.map(x => this.padding(x, length)).join("")
        }
        return frags.slice(0, size).map(x => this.padding(x, length)).join("")
    }

    private static reduce(v: string, length: number): string {
        if (!v) {
            return v;
        }
        const bu: string[] = [];
        for (let index = 0; index < v.length; index += length) {
            let hasNoneZero: boolean = true;
            for (let offset = index; offset < index + length; offset++) {
                if (v.charAt(offset) != '0') {
                    hasNoneZero = false;
                }
                if (!hasNoneZero) {
                    bu.push(v.charAt(offset));
                }
            }
            if (hasNoneZero) {
                bu.push('0');
            }
            bu.push('.');
        }
        return bu.slice(0, bu.length - 1).join("");
    }
}